### PR TITLE
Adding new "Add To Top" feature for Textarea Choice Lists

### PR DESCRIPTION
**Pull request by derFunk**
There's now a checkbox named "Add To Top" for Textarea Choice Lists which allow for adding new user-added values on top of the combo box list.

This is handy if you want to have all user-added values as a default value for upcoming builds. For that to work you'd have to set "Default Choice" to "Top Most Value".

image

We're using this change already in production, yet:
**Disclaimer: There are no specific Unit Tests for this feature added yet! I'm asking somebody else to add them, since the Unit Test setup for this plugin is not working for me currently.**.
See also https://issues.jenkins-ci.org/browse/JENKINS-34371.

### CODE REVIEW

File: src/main/java/jp/ikedam/jenkins/plugins/extensible_choice_parameter/TextareaChoiceListProvider.java [hunk 1]
```diff
@@ -143,6 +143,21 @@ public String getDefaultChoice()
...
+     * 
+     * @return if user-added values should be added to the top of the choice list.
+     * @see jp.ikedam.jenkins.plugins.extensible_choice_parameter.ChoiceListProvider#isAddToTop()
+     */
+    @Override
+    public boolean isAddToTop()
+    {
+        return addToTop;
+    }
+
```
Comment (ikedam) [inline src/main/java/jp/ikedam/jenkins/plugins/extensible_choice_parameter/TextareaChoiceListProvider.java]: Move to `AddEditedChoiceListProvider`.
This should be treated like `whenToAdd`.

File: src/main/java/jp/ikedam/jenkins/plugins/extensible_choice_parameter/ChoiceListProvider.java [hunk 2]
```diff
@@ -68,6 +68,16 @@ public String getDefaultChoice()
...
+     * Returns if user-added values should be added to the top of the choice list.
+     * 
+     * @return if user-added values should be added to the top of the choice list.
+     */
+    public boolean isAddToTop()
+    {
+        return false;
+    }
+    
+    /**
```
Comment (ikedam) [inline src/main/java/jp/ikedam/jenkins/plugins/extensible_choice_parameter/ChoiceListProvider.java]: This should be applied to `AddEditedChoiceListProvider`.

File: src/main/java/jp/ikedam/jenkins/plugins/extensible_choice_parameter/TextareaChoiceListProvider.java [hunk 3]
```diff
@@ -154,13 +169,15 @@ public String getDefaultChoice()
      * @param defaultChoice
      * @param addEditedValue
      * @param whenToAdd
+     * @param addToTop
      */
     @DataBoundConstructor
-    public TextareaChoiceListProvider(String choiceListText, String defaultChoice, boolean addEditedValue, WhenToAdd whenToAdd)
+    public TextareaChoiceListProvider(String choiceListText, String defaultChoice, boolean addEditedValue, WhenToAdd whenToAdd, boolean addToTop)
```
Comment (ikedam) [inline src/main/java/jp/ikedam/jenkins/plugins/extensible_choice_parameter/TextareaChoiceListProvider.java]: Would you add a backward compatible constructor?
example: https://github.com/jenkinsci/naginator-plugin/blob/naginator-1.17/src/main/java/com/chikli/hudson/plugin/naginator/NaginatorPublisher.java#L53

Adding  a backward compatible constructor preserves binary compatibilities,
and you no longer need to edit existing test codes.
(I didn't add backward compatible constructor in previous changes for extebsible-choice-parameter, and I know that was a really bad idea...)

File: src/main/java/jp/ikedam/jenkins/plugins/extensible_choice_parameter/TextareaChoiceListProvider.java [hunk 4]
```diff
@@ -172,8 +189,15 @@ protected void addEditedValue(
...
         LOGGER.info(String.format("Add new value %s to parameter %s in project %s", value, def.getName(), project.getName()));
         List<String> newChoiceList = new ArrayList<String>(getChoiceList());
-        newChoiceList.add(value);
+        
+        // check if the new value should be added to the top of the list, or not
+        if (this.addToTop)
+        	newChoiceList.add(0, value);
+        else
+        	newChoiceList.add(value);
+
```
Comment (ikedam) [inline src/main/java/jp/ikedam/jenkins/plugins/extensible_choice_parameter/TextareaChoiceListProvider.java]: 1. There're tab characters. Please use white spaces for indentations.
2. I highly recommend you to use braces ({}) even for single statements. Blocks without braces often causes bugs especially when other developer add changes to your code.

File: src/main/resources/jp/ikedam/jenkins/plugins/extensible_choice_parameter/taglib/addEditedValue.jelly [hunk 5]
```diff
@@ -30,5 +30,8 @@ THE SOFTWARE.
     <f:entry title="${%When to Add}" field="whenToAdd" help="/plugin/extensible-choice-parameter/help/whenToAdd.html">
         <f:enum>${it}</f:enum>
     </f:entry>
+	<f:entry title="${%Add To Top}" field="addToTop" help="/plugin/extensible-choice-parameter/help/addToTop.html">
```
Comment (ikedam) [inline src/main/resources/jp/ikedam/jenkins/plugins/extensible_choice_parameter/taglib/addEditedValue.jelly]: A tab character.

File: src/main/webapp/help/addToTop.html [hunk 6]
```diff
@@ -0,0 +1,4 @@
+<div>
+If checked, the user-added value will be added on top of the combo box, making it the new default value for the next build, 
+if <i>Default Choice</i> is set to <i>Top Most Value</i>.
+</div>
```
Comment (ikedam) [inline src/main/webapp/help/addToTop.html]: JFYI: It's better to have text files always end with new line characters.
(Some text editors automatically add new lines at the ends)

**Comment by ikedam** [comment]
The logic looks good to me.
Also should be applied to `GlobalTextareaChoiceListProvider`

**Comment by derFunk** [comment]
@ikedam I implemented your suggestions :+1: Didn't apply to GlobalTextareaChoiceListProvider yet though.

**Comment by ikedam** [comment]
The update looks good to me.
I'm looking for the next update applying to `GlobalTextareaChoiceListProvider` and adding tests.

**Comment by ikedam** [comment]
~~I'm looking for the next update~~
I'm looking forward to the next update

**Pull request merged**